net: shell: Add priority to ping

krish2718 opened the pull request:
This is handy in testing of setting priority directly rather than deriving from DSCP. Please note ICMP doesn't use net context.

krish2718 committed:
net: Add priority to ping

This is handy in testing of setting priority directly rather than
deriving from DSCP. Please note ICMP doesn't use net context.

This is applicable for both shell and API.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

pdgendt commented on subsys/net/lib/zperf/zperf_shell.c:
```diff
@@ -526,7 +526,7 @@ static int execute_upload(const struct shell *sh,
 		 */
 		net_icmpv6_send_echo_request(net_if_get_default(),
-					     &ipv6->sin6_addr, 0, 0, 0, NULL, 0);
+					     &ipv6->sin6_addr, 0, 0, 0, 0, NULL, 0);
```
Use `-1` to keep the current behavior?

pdgendt commented on subsys/net/ip/icmpv6.c:
```diff
@@ -362,8 +363,12 @@ int net_icmpv6_send_echo_request(struct net_if *iface,
 
-	net_pkt_set_ip_dscp(pkt, net_ipv6_get_dscp(tc));
-	net_pkt_set_ip_ecn(pkt, net_ipv6_get_ecn(tc));
+	if (priority == -1) {
```
```suggestion
	if (priority >= NET_MAX_PRIORITIES) {
		return -EINVAL;
	}
	if (priority < 0) {
```

krish2718 replied:
Well, the use case that I am using needs any arbitrary priority to be assigned to the packet, and in all places the if the priority is greater than suported, it falls back to minimal one. e.g., `net_tx_priority2tc`, so, should we allow any priority from the shell?

jukkar replied:
> should we allow any priority from the shell

IMHO we should not allow any arbitrary values to be used here by default. The shell can be used for development purposes but it is mainly meant for end users. If we really want some way to set special values, then I suggest a Kconfig option that is by default 'n' and which enables such behavior.

krish2718 replied:
I am fine with adding a Kconfig option `NET_ALLOW_ANY_PRIORITY` that is disabled by default.

jukkar replied:
Sounds good to me.

krish2718 committed:
net: Add configuration option to allow any priority

This adds support to allow any priority from the user instead of
limiting to the protocol values (0-7). This is useful in conveying
custom priorities from application to the driver/chipset.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
zperf: Add support to configure context priority

This is handy in setting a custom priority for the context either
through shell or API.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

pdgendt commented on subsys/net/ip/icmpv4.c:
```diff
@@ -537,8 +538,19 @@ int net_icmpv4_send_echo_request(struct net_if *iface,
 
-	net_pkt_set_ip_dscp(pkt, net_ipv4_get_dscp(tos));
-	net_pkt_set_ip_ecn(pkt, net_ipv4_get_ecn(tos));
+#if !defined(CONFIG_NET_ALLOW_ANY_PRIORITY)
+	if (priority > NET_MAX_PRIORITIES) {
+		NET_ERR("Priority %d is too large, maximum is %d",
+			priority, NET_MAX_PRIORITIES);
+		return -EINVAL;
+	}
+#endif /* !CONFIG_NET_ALLOW_ANY_PRIORITY */
```
The priority can't be equal to `NET_MAX_PRIORITIES` (same for `IPv6` and `zperf`).

Nits: Use `#ifndef` and add a newline after `#endif`.

```suggestion
#ifndef CONFIG_NET_ALLOW_ANY_PRIORITY
	if (priority >= NET_MAX_PRIORITIES) {
		NET_ERR("Priority %d is too large, maximum is %d",
			priority, NET_MAX_PRIORITIES);
		return -EINVAL;
	}
#endif /* !CONFIG_NET_ALLOW_ANY_PRIORITY */

```

jukkar replied:
I think we should print the max like this

`NET_ERR("Priority %d is too large, maximum is %d", priority, NET_MAX_PRIORITIES - 1);`

if we check `priority >= NET_MAX_PRIORITIES`

jukkar commented on subsys/net/ip/Kconfig:
```diff
@@ -290,6 +290,12 @@ config NET_RX_DEFAULT_PRIORITY
 	  one. The value 0 means lowest priority and 7 is the highest.
 
+config NET_ALLOW_ANY_PRIORITY
+	bool "Allow any priority to be used"
+	help
+	  If this is set, then any user given priority can be used. Otherwise the
+	  priorities are limited to 0-7 range.
```
I think we need to emphasize that the priority is for network packets, it is not obvious now from the text.
So something like
```
bool "Allow any network packet priority to be used"
help
   If this is set, then any user given network packet priority can be used. Otherwise the
   network packet priorities are limited to 0-7 range.
```

jukkar commented on subsys/net/ip/icmpv6.c:
```diff
@@ -363,11 +363,13 @@ int net_icmpv6_send_echo_request(struct net_if *iface,
 		return -ENOMEM;
 	}
 
+#if !defined(CONFIG_NET_ALLOW_ANY_PRIORITY)
```
Same here

jukkar reviewed (changes requested):
Please submit a version that fixes all the comments, not just some of them.

krish2718 commented:
> Please submit a version that fixes all the comments, not just some of them.

Sorry, was looking at my e-mail not the PR, missed one review.

krish2718 commented:
> > Please submit a version that fixes all the comments, not just some of them.
> 
> Sorry, was looking at my e-mail not the PR, missed one review.

Well, its not that, I had resolved all the comments, but messed up the rebase :-(. Its now fixed.

jukkar reviewed (approved):
LGTM
